### PR TITLE
Support multiple instance

### PR DESCRIPTION
**Pull request by Lbqds**

### CODE REVIEW

**Comment by killerwhile** [comment]
LGMT, thanks!

File: test/shareProcessorTest.js [hunk 1]
```diff
@@ -26,6 +26,49 @@ describe('test share processor', function(){
         })
     })
 
+    it('should ignore duplicated shares', function(done){
+        var shareProcessor = new ShareProcessor(test.config, test.logger);
+        shareProcessor.redisClient = redisClient;
+
+        var share1 = {job: {fromGroup: 0, toGroup: 1}, foundBlock: false, blockHash: 'hash1', difficulty: 1, workerAddress: 'miner1'};
+        var share2 = {job: {fromGroup: 0, toGroup: 1}, foundBlock: false, blockHash: 'hash2', difficulty: 2, workerAddress: 'miner1'};
+        var invalidShare = {job: {fromGroup: 0, toGroup: 1}, foundBlock: false, blockHash: 'hash1', difficulty: 3, workerAddress: 'miner1'};
```
Comment (killerwhile) [inline test/shareProcessorTest.js]: Shouldn't it more relevant to be from a different worker address?

File: lib/shareProcessor.js [hunk 2]
```diff
@@ -86,33 +86,56 @@ var ShareProcessor = module.exports = function ShareProcessor(config, logger){
...
+
+            var currentTs = Math.floor(currentMs / 1000);
+            redisTx.zadd(hashrateKey, currentTs, [fromGroup, toGroup, share.worker, share.difficulty, currentMs].join(':'));
+
+            if (share.foundBlock){
+                var blockHash = share.blockHash;
+                var newKey = _this.roundKey(fromGroup, toGroup, blockHash);
+                var blockWithTs = blockHash + ':' + currentMs.toString();
+
+                redisTx.rename(currentRound, newKey);
```
Comment (killerwhile) [inline lib/shareProcessor.js:131]: `renamenx` should probably be used here, otherwise in case of found block sent with delay (but still not expired) by one worker, the `sharecache` key would have been deleted and the shares for the block overwritten by new currentRound.

File: lib/shareProcessor.js [hunk 3]
```diff
@@ -86,33 +86,56 @@ var ShareProcessor = module.exports = function ShareProcessor(config, logger){
...
+                redisTx.sadd(pendingBlocksKey, blockWithTs);
+                redisTx.hset(foundBlocksKey, blockHash, share.workerAddress)
```
Comment (killerwhile) [inline lib/shareProcessor.js:133]: `hsetnx` should be used here, so that "paternity" of the block is kept to the first worker who submitted it (in case of same corner case than above)

**Review by killerwhile — CHANGES_REQUESTED** [review]
I reviewed more in depth the atomicity logic of the duplicate solution corner case, and it could be improved.

File: test/shareProcessorTest.js (hunk 1 above)
Comment (Lbqds) [reply test/shareProcessorTest.js]: Now we check if a share is duplicated based on hash, but it's good for testing, thanks.

File: lib/shareProcessor.js (hunk 2 above)
Comment (Lbqds) [reply lib/shareProcessor.js:131]: Great catch, thanks, the remaining commands will continue to execute even if `renamenx` failed in redis transaction, so my thought is that we can use `set hash true ex expiryPeriod nx` to check if the share is duplicated, which `expiryPeriod > jobExpiryPeriod`, how do you think?

File: lib/shareProcessor.js [hunk 4]
```diff
@@ -86,33 +86,56 @@ var ShareProcessor = module.exports = function ShareProcessor(config, logger){
...
+                redisTx.del(cacheKey);
```
Comment (killerwhile) [inline lib/shareProcessor.js:133]: the cacheKey is only removed when a corresponding block is found. Since it's not possible to set TTL on set, we should have a cleanup mechanism which empty this set, otherwise it might quickly become unnecessarily big.

**Review by rfkih — APPROVED** [review]
hello

**Comment by sergeybrava** [comment]
-_-
